2020/5/boarding_passes/src/main.rs

```diff
@@ -0,0 +1,168 @@
+use std::collections::HashSet;
+use std::env;
+use std::fs::File;
+use std::io::{self, BufRead};
+
+fn main() -> Result<(), Error> {
+    let args: Vec<String> = env::args().collect();
+    let file = File::open(&args[1])?;
+    let reader = io::BufReader::new(file);
+
+    // if let Ok(max) = max_id(reader) {
+    //     println!("The largest id is {}", max);
+    // }
+
+    let mut all_passes = HashSet::new();
+    for line in reader.lines() {
+        if let Ok(l) = line {
+            let bp = BoardingPass::new(&l)?;
+            all_passes.insert(bp.seat_id());
+        }
+    }
+
+    for row in 1..127 {
+        for col in 0..7 {
+            let id = 8 * row + col;
+            if !all_passes.contains(&id)
+                && all_passes.contains(&(id - 1))
+                && all_passes.contains(&(id + 1))
+            {
+                println!("Your seat id is {}", id);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn max_id(reader: io::BufReader<std::fs::File>) -> Result<usize, Error> {
```

Comment:
Was considering making this generic taking `io::BufReader<T>`, but that let to trait errors that weren't clear to me how to specify the right constraint for T :(

Reply:
Since `lines` is on `BufRead` (the trait that `BufReader` implements) you could just do something like

```rust
fn max_ids<R: BufRead>(reader: R) ...
```

Reply:
Ah, that would make sense. Thanks!